### PR TITLE
Delete unnecessary separators and don't show menu when there are no visible items

### PR DESCRIPTION
**Pull request by serjN**
Delete separators when previous elements are not visible and doesn't show menu when no visible elements are available.

### CODE REVIEW

**Comment by sindresorhus** [comment]
Your time would be better spent fixing the actual issue https://github.com/electron/electron/issues/5869 so everyone would benefit, not just users of this module.

**Comment by serjN** [comment]
OK, I will look at it later. But now I think this pull request is useful for this module.

File: index.js [hunk 1]
```diff
@@ -89,13 +89,28 @@ function create(win, opts) {
 
 		// filter out leading/trailing separators
 		// TODO: https://github.com/electron/electron/issues/5869
-		menuTpl = menuTpl.filter((el, i, arr) => !(el.type === 'separator' && (i === 0 || i === arr.length - 1 || arr[i + 1].type === 'separator')));
+		menuTpl = deleteUnnecessarySeparatorsAndElements(menuTpl);
 
-		const menu = (electron.Menu || electron.remote.Menu).buildFromTemplate(menuTpl);
-		menu.popup(win);
+		if (menuTpl.some(elem => isVisible(elem))) {
```
Comment (sindresorhus) [inline index.js]: This menu is already filtered, so couldn't you just do: `if (menuTpl.length > 0) {` ?

File: index.js [hunk 2]
```diff
@@ -89,13 +89,28 @@ function create(win, opts) {
...
+	let visiblePreviousEl;
+	return menuTpl.filter(el => isVisible(el)).filter((el, i, arr) => {
+		const toDelete = el.type === 'separator' && (!visiblePreviousEl || i === arr.length - 1 || arr[i + 1].type === 'separator');
+		visiblePreviousEl = toDelete || !isVisible(el) ? visiblePreviousEl : el;
+		return !toDelete;
 	});
 }
 
+function isVisible(elem) {
+	return !('visible' in elem) || elem.visible;
```
Comment (sindresorhus) [inline index.js]: Could be simplified to just `return elem.visible !== false;`

File: index.js [hunk 3]
```diff
@@ -89,13 +89,28 @@ function create(win, opts) {
...
+function deleteUnnecessarySeparatorsAndElements(menuTpl) {
+	let visiblePreviousEl;
+	return menuTpl.filter(el => isVisible(el)).filter((el, i, arr) => {
+		const toDelete = el.type === 'separator' && (!visiblePreviousEl || i === arr.length - 1 || arr[i + 1].type === 'separator');
+		visiblePreviousEl = toDelete || !isVisible(el) ? visiblePreviousEl : el;
+		return !toDelete;
 	});
 }
 
+function isVisible(elem) {
```
Comment (sindresorhus) [inline index.js]: Just put this as an arrow function inside `deleteUnnecessarySeparatorsAndElements`.

File: index.js [hunk 4]
```diff
@@ -89,13 +89,28 @@ function create(win, opts) {
...
-		const menu = (electron.Menu || electron.remote.Menu).buildFromTemplate(menuTpl);
-		menu.popup(win);
+		if (menuTpl.some(elem => isVisible(elem))) {
+			const menu = (electron.Menu || electron.remote.Menu).buildFromTemplate(menuTpl);
+			menu.popup(win);
+		}
+	});
+}
+
+function deleteUnnecessarySeparatorsAndElements(menuTpl) {
```
Comment (sindresorhus) [inline index.js]: Needs a shorter name.

File: index.js [hunk 5]
```diff
@@ -89,13 +89,28 @@ function create(win, opts) {
...
+	let visiblePreviousEl;
+	return menuTpl.filter(el => isVisible(el)).filter((el, i, arr) => {
+		const toDelete = el.type === 'separator' && (!visiblePreviousEl || i === arr.length - 1 || arr[i + 1].type === 'separator');
+		visiblePreviousEl = toDelete || !isVisible(el) ? visiblePreviousEl : el;
```
Comment (sindresorhus) [inline index.js]: Why are you doing `!isVisible(el)` check? You already filtered out all invisible elements on line 103.

**Comment by sindresorhus** [comment]
@serjN Can you finish this up?

**Comment by serjN** [comment]
I will finish on this week.

**Pull request merged**